Service/ExecuteSourceDeductionForItems.php

```diff
@@ -0,0 +1,148 @@
+     */
+    public function __construct(
+        WebsiteRepositoryInterface $websiteRepository,
+        SalesEventInterfaceFactory $salesEventFactory,
+        SalesChannelInterfaceFactory $salesChannelFactory,
+        ItemToDeductFactory $itemToDeductFactory,
+        SourceDeductionRequestFactory $sourceDeductionRequestFactory,
+        SourceDeductionService $sourceDeductionService,
+        SourcesRepositoryInterface $sourceRepository,
+        Processor $priceIndexer
+    ) {
+        $this->websiteRepository = $websiteRepository;
+        $this->salesEventFactory = $salesEventFactory;
+        $this->salesChannelFactory = $salesChannelFactory;
+        $this->itemToDeductFactory = $itemToDeductFactory;
+        $this->sourceDeductionRequestFactory = $sourceDeductionRequestFactory;
+        $this->sourceDeductionService = $sourceDeductionService;
+        $this->sourceRepository = $sourceRepository;
+        $this->priceIndexer = $priceIndexer;
+    }
+
+    /**
+     * @param OrderInterface $order
+     * @param GetItemsToCancelFromOrderItem[] $itemsToCancel
+     * @param Creditmemo|null $creditmemo
+     */
+    public function executeSourceDeductionForItems(OrderInterface $order, array $itemsToCancel, Creditmemo $creditmemo = null)
+    {
+        $type = SalesEventInterface::EVENT_ORDER_CANCELED;
+        if ($creditmemo) {
+            $type = SalesEventInterface::EVENT_CREDITMEMO_CREATED;
+        }
+
+        $websiteId = $order->getStore()->getWebsiteId();
+        $websiteCode = $this->websiteRepository->getById($websiteId)->getCode();
+        $salesChannel = $this->salesChannelFactory->create([
+            'data' => [
+                'type' => SalesChannelInterface::TYPE_WEBSITE,
+                'code' => $websiteCode
+            ]
+        ]);
+
+        $salesEvent = $this->salesEventFactory->create([
+            'type' => $type,
+            'objectType' => SalesEventInterface::OBJECT_TYPE_ORDER,
+            'objectId' => (string)$order->getEntityId()
+        ]);
+
+        foreach ($itemsToCancel as $item) {
+            if ($creditmemo && !$item->getBackToStock()) {
+                continue;
+            }
+
+            $sourceItem = $this->sourceRepository->getSourceItemBySku(
+                (string)$order->getEntityId(), $item->getSku()
+            );
+
+            $sourceCode = $sourceItem ? $sourceItem->getSourceCode() : 'default';
+
+            $sourceDeductionRequest = $this->sourceDeductionRequestFactory->create([
+                'sourceCode' => $sourceCode,
+                'items' => [$this->itemToDeductFactory->create([
+                    'sku' => $item->getSku(),
+                    'qty' => -$item->getQty()
+                ])],
+                'salesChannel' => $salesChannel,
+                'salesEvent' => $salesEvent
+            ]);
+
+            $this->sourceDeductionService->execute($sourceDeductionRequest);
+            $this->priceIndexer->reindexRow($item->getProductId());
```

Comment:
Maybe have a look at the `reindexList` method here. The method takes an array of IDs rather than a single one. If we build an array of IDs to pass to the function then we'd only query MySQL once, rather than every time we iterate `$itemsToCancel`